### PR TITLE
Improve `COPY FROM` retry logic

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Resolves https://github.com/crate/crate/issues/13110

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 1]
```diff
@@ -256,16 +261,20 @@ private String getLine(BufferedReader reader, long startFrom, int retry) throws
...
-        } catch (SocketTimeoutException e) {
-            if (retry > MAX_SOCKET_TIMEOUT_RETRIES) {
-                URI uri = currentInput.uri();
-                LOGGER.error("Timeout during COPY FROM '" + uri.toString() + "' after " + retry + " retries", e);
-                throw e;
-            } else {
+        } catch (SocketException | SocketTimeoutException e) {
+            if (backoffPolicy.hasNext()) {
+                try {
+                    Thread.sleep(backoffPolicy.next().getMillis());
```
Comment (seut) [inline server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:267]: Could we use a scheduler instead to not block this thread?

Comment (jeeminso) [reply server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:267]: Sorry I am confused how to use a scheduler within a `BatchIterator`. Could we have quick chat about it?
Also, how about we let `BatchIteratorBackpressureExecutor` to account for delays from `FileReadingIterator.moveNext()`?
https://github.com/crate/crate/blob/ac60750e1e651b563f5ed14a94c97978ae860287/server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java#L169

Comment (jeeminso) [reply server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: Could you have another look at this draft? @seut

**Comment by matriv** [comment]
Please bear with me If I miss something in the context.

1. The `FileReadingIterator` now, doesn't handle the Socket exceptions but instead rethrows the Exception and closes the reader? is this correct?
2. The `BatchIteratorBackpressureExecutor` which wraps the `FileReadingIterator`, now handles the Socket exceptions and will resume the execution after delay (given by the policy), but now the reader is already closed?
3. If this is the way to go, is there a way unify the `getDelayInMs` mechanism with the backoff policy (in case of socket exception)?

**Comment by jeeminso** [comment]
Thanks for reviewing @matriv,

>The FileReadingIterator now, doesn't handle the Socket exceptions but instead rethrows the Exception and closes the reader? is this correct?
Yes that is correct (but not just closes, closes and re-initializes, which were existing logic).
>The BatchIteratorBackpressureExecutor which wraps the FileReadingIterator, now handles the Socket exceptions and will resume the execution after delay (given by the policy), but now the reader is already closed?
Yes that is also correct but the reader is re-initialized such that another invocation of `moveNext()` would work smoothly.
>If this is the way to go, is there a way unify the getDelayInMs mechanism with the backoff policy (in case of socket exception)?
So far was focused more on implementing non-blocking back-off policy. I will look into this Thank you.

**Comment by matriv** [comment]
> Yes that is correct (but not just closes, closes and re-initializes, which were existing logic).

Apologies, I missed this.

 
> Yes that is also correct but the reader is re-initialized such that another invocation of `moveNext()` can resume smoothly.

ok. thx!


I don't like that much (design-wise) that the `BatchIteratorBackpressureExecutor` needs to know about the Socket exception, and implement some concrete logic, although I don't have a concrete idea to somehow make this more generic in an elegant way.
Is there a way for the `ShardUpsertExecutor` to know that it's being called by `COPY FROM` so we can create some new class extending the `BatchIteratorBackpressureExecutor`, which also handles those SocketExceptions with a backoff policy?
Please wait for @seut review, as I'm not so familiar, and don't want to confuse/de-rail you.

**Comment by seut** [comment]
@jeeminso 
 
> > The FileReadingIterator now, doesn't handle the Socket exceptions but instead rethrows the Exception and closes the reader? is this correct?
> 
> Yes that is correct (but not just closes, closes and re-initializes, which were existing logic).
> 
> > The BatchIteratorBackpressureExecutor which wraps the FileReadingIterator, now handles the Socket exceptions and will resume the execution after delay (given by the policy), but now the reader is already closed?
> 
> Yes that is also correct but the reader is re-initialized such that another invocation of `moveNext()` can resume smoothly.
> 

Is this really working as expected given that both exception the BatchIteratorBackpressureExecutor should handle are IOExceptions which are catched by `FileReadingIterator.moveNext()` at https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java#L187 ?

Is your draft tested/validated somehow? If not, this would help validating your approach. 

Otherwise the solution looks good to me.

**Comment by seut** [comment]
> I don't like that much (design-wise) that the `BatchIteratorBackpressureExecutor` needs to know about the Socket exception, and implement some concrete logic, although I don't have a concrete idea to somehow make this more generic in an elegant way.
> Is there a way for the `ShardUpsertExecutor` to know that it's being called by `COPY FROM` so we can create some new class extending the `BatchIteratorBackpressureExecutor`, which also handles those SocketExceptions with a backoff policy?

I wonder if may could/should catch generic `IOException`s instead, this would also make it more generic and not bound to socket errors. @mfussenegger wdyt?

**Review by mfussenegger — COMMENTED** [review]
I'm not sure about this approach. So far the `BackpressureExecutor` was reacting to backpressure applied from the write *target*. 

With this change it would gain awareness of exceptions raised by the *source* and handle that too.

We could parameterize the classes, but even then I don't really like the tight coupling of the components.

To *read* from external sources resiliently we would depend on a concrete other component being used.

Can't we use the `loadNextBatch` mechanism of the BatchIterator to keep this contained to the `FileReadingBatchIterator` ?

In the future we may also want to change the `FileReadingIterator` to use async APIs for sources like S3, and then we'd need some way to have non-blocking operations anyways.

**Comment by mfussenegger** [comment]
Another approach that came to mind is to have a dedicated `RetryBatchIterator(BatchIterator<T> bi, Iterable<TimeValue> backoffPolicy, List<Class<?>> retryExceptions)`. (Or `Predicate<Throwable>`).

That would allow to add retry handling on top of any batchIterator. The advantage over extending `BackpressureExecutor` is that it doesn't change the interface, but is transparent to consumers. And it could be used independently. It's probably easier to test as well.

**Comment by jeeminso** [comment]
Hi @mfussenegger I have [5aa4b95](https://github.com/crate/crate/pull/13749/commits/5aa4b9547199f1f742e835beb053af352b126ab0) that is a rough draft of `RetryBatchIterator` but
> That would allow to add retry handling on top of any batchIterator.

seems not possible since the source batchIterator is required to catch the exceptions and get ready to re-try.
For example, a source batchIterator may react to an exception by closing/releasing resources or subsequent calls to `moveNext()` may not resume from expected index.

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 2]
```diff
@@ -292,13 +314,20 @@ private void releaseBatchIteratorState() {
...
+            throw lastSocketException;
+        }
+        return cf;
     }
 
     @Override
     public boolean allLoaded() {
-        return true;
+        // always allow to invoke loadNextBatch()
+        return false;
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: Yes, Thanks!

**Comment by jeeminso** [comment]
- status update:
Taking `loadNextBatch()` [approach](https://github.com/crate/crate/pull/13749#pullrequestreview-1339536835).
The draft implements retry with back-off policy and the only thing remaining should be testing. Currently we have a single test covering retries caused by socket exceptions, I will put a few more tests with socket exceptions combined with headers and multiple URIs, etc.

**Comment by jeeminso** [comment]
Hello @seut, it would be great if you could have a quick look at my new draft when you have time. Thanks!

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingCollectorTest.java [hunk 3]
```diff
@@ -71,14 +74,24 @@
...
     private static File tmpFileGz;
     private static File tmpFileEmptyLine;
     private InputFactory inputFactory;
     private Input<String> sourceUriFailureInput;
     private TransactionContext txnCtx = CoordinatorTxnCtx.systemTransactionContext();
 
+    @BeforeClass
+    public static void setupThreadPool() {
+        THREAD_POOL = new TestThreadPool(Thread.currentThread().getName());
+    }
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingCollectorTest.java:88]: Isn't it flaky to rely on that this is called before `setUpClass`? why not do this inside `setUpClass`?

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 4]
```diff
@@ -243,30 +266,33 @@ private void closeCurrentReader() {
...
                 closeCurrentReader();
                 initCurrentReader(currentInput, currentUri);
-                return getLine(currentReader, startLine, retry + 1);
+            } else {
+                URI uri = currentInput.uri();
+                LOGGER.error("Timeout during COPY FROM '" + uri.toString() +
+                             "' after " + MAX_SOCKET_TIMEOUT_RETRIES +
+                             " retries", e);
             }
+            throw e;
```
Comment (seut) [inline server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:295]: Wouldn't it be cheaper while still valid to utilize a `NULL` return value as the indication for returning `FALSE` at `moveNext()`?  re-throwing and catching exceptions may be an expensive operation due to the stack trace frames...

**Comment by seut** [comment]
@jeeminso I think the approach looks good.

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingCollectorTest.java (hunk 3 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/execution/engine/collect/files/FileReadingCollectorTest.java]: Thank you, done

**Comment by jeeminso** [comment]
retest this please: `test_copy_from_with_fail_fast_with_single_node`
```
java.lang.AssertionError: 
Expecting actual throwable to be exactly an instance of:
  io.crate.exceptions.JobKilledException
but was:
  java.lang.IllegalStateException: BatchIterator is closed
	at io.crate.data.BatchIterator.<clinit>(BatchIterator.java:73)
	at io.crate.testing.TestingBatchIterators.range(TestingBatchIterators.java:41)
	at io.crate.action.sql.CursorTest.rows(CursorTest.java:640)
	...(41 remaining lines not displayed - this can be changed with Assertions.setMaxStackTraceElementsDisplayed)
```
I don't think the current changes will cause this flakyness which I cannot reproduce with several repetitions locally.

**Review by seut — COMMENTED** [review]
Just missing a Changes entry, otherwise looks good!

File: docs/appendices/release-notes/unreleased.rst [hunk 5]
```diff
@@ -64,6 +64,10 @@ Changes
 - Added support for ``ORDER BY``, ``MAX``, ``MIN`` and comparison operators on
   expressions of type ``INTERVAL``.
 
+- Added a back-off policy to ``COPY FROM`` retry logic to pause temporarily
+  in exponential increments between retries caused by ``SocketExceptions`` and
+  ``SocketTimeoutExceptions``.
```
Comment (seut) [inline docs/appendices/release-notes/unreleased.rst:69]: I'd phrase it more user facing than implementation wise:
```suggestion
- Improved ``COPY FROM`` retry logic to retry exponentially on temporary
  network timeout and general network errors.
```

Comment (jeeminso) [reply docs/appendices/release-notes/unreleased.rst]: Thank you! updated.

File: docs/appendices/release-notes/unreleased.rst [hunk 6]
```diff
@@ -64,6 +64,9 @@ Changes
 - Added support for ``ORDER BY``, ``MAX``, ``MIN`` and comparison operators on
   expressions of type ``INTERVAL``.
 
+- Improved ``COPY FROM`` retry logic to retry exponentially on temporary
```
Comment (matriv) [inline docs/appendices/release-notes/unreleased.rst:67]:
```suggestion
- Improved ``COPY FROM`` retry logic to retry with a delay which increases exponentially on temporary
```

**Comment by jeeminso** [comment]
Is the current back-off delays good enough?
`BackoffPolicy.exponentialBackoff(TimeValue.ZERO, MAX_SOCKET_TIMEOUT_RETRIES)`

**Comment by seut** [comment]
> Is the current back-off delays good enough? `BackoffPolicy.exponentialBackoff(TimeValue.ZERO, MAX_SOCKET_TIMEOUT_RETRIES)`

Yes, we could add 10ms initial delay, but we also retried immediately with the old code. I think it's fine to do the first retry immediately.

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java (hunk 4 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:290]: Just found out that the following change:
```
catch (IOException e) {
    if ((e instanceof SocketException || e instanceof SocketTimeoutException) && backOffPolicy.hasNext()) {
        return false;
    }
```
catches socket exceptions thrown while reading headers or [skipping lines](https://crate.io/docs/crate/reference/en/5.3/sql/statements/copy-from.html#skip) which makes this feature overlap with https://github.com/crate/crate/issues/13276. 
Unless we handle https://github.com/crate/crate/issues/13276 we should revert this change. What do you think? @seut

**Review by seut — APPROVED** [review]
ah good catch, so lets do it like that now, don't have a better suggestion atm

**Pull request merged**